library/core/src/ptr/mod.rs

```diff
@@ -55,20 +54,46 @@
 //! has size 0, i.e., even if memory is not actually touched. Consider using
 //! [`NonNull::dangling`] in such cases.
 //!
-//! ## Allocated object
+//! ## Allocated Object and Provenance
 //!
 //! For several operations, such as [`offset`] or field projections (`expr.field`), the notion of an
 //! "allocated object" becomes relevant. An allocated object is a contiguous region of memory.
 //! Common examples of allocated objects include stack-allocated variables (each variable is a
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//! When an object is allocated there should be only one way to access to it (the variable's name,
+//! the pointer returned by malloc). This One True Handle is given a unique *provenance* which
+//! gives it permission to access that object. This provenance, and therefore permission to access
+//! the allocation, is implicitly shared with all pointers that are either directly or transitively
+//! *derived* from the One True Handle through operations like `offset` or borrowing.
+//!
+//! (Unclear detail: taking a subslice is proposed to create a slice that is no longer allowed
+//! to access the full range of memory -- is this part of provenance or another system?)
```

Comment:
I view this as part of provenance. With Stacked Borrows, provenance is more refined than just an allocation ID; that's what the "tags" are for. (And technically we probably would not even need allocation IDs any more; Stacked Borrows tags could be our only provenance.)